### PR TITLE
Linting and Component Refactoring

### PR DESCRIPTION
**Pull request by theduckylittle**
- Added new stricter linting rules.
- Cleaned up code based on those rules.
- Refactoring components to provide clearer division of labor (State, Rendering, Logic are now all more clearly divided).
- Ensured that `const` is used appropriately for better browser hinting.
- Replacing `for(... in ...)` with proper `for(... i < ii ...)` loops as the `for .. in` loops are transformed poorly and performance is bad.

### CODE REVIEW

File: src/gm3/components/catalog/legend.js [hunk 1]
```diff
@@ -1,10 +1,10 @@
 /*
  * The MIT License (MIT)
  *
- * Copyright (c) 2016-2017 Dan "Ducky" Little
+ * Copyright (c) 2016-2017 Dan 'Ducky' Little
  *
  * Permission is hereby granted, free of charge, to any person obtaining a copy
- * of this software and associated documentation files (the "Software"), to deal
+ * of this software and associated documentation files (the 'Software'), to deal
```
Comment (klassenjs) [inline src/gm3/components/catalog/legend.js]: Why do some of these use ' and others use "  ?

File: src/gm3/actions/mapSource.js [hunk 2]
```diff
@@ -658,9 +657,9 @@ export function changeFeatures(mapSourceName, filter, properties) {
  *
  */
 export function getOrderedMapSources(mapSources) {
-    let active = [];
-    for(let ms in mapSources) {
-        if(isMapSourceActive(map_sources[ms])) {
+    const active = [];
```
Comment (klassenjs) [inline src/gm3/actions/mapSource.js]: How did this work before?

**Comment by klassenjs** [comment]
npm install is failing for me.  Any ideas?

```npm ERR! write after end
npm ERR! write after end
npm ERR! write after end
npm ERR! write after end
npm ERR! write after end
npm ERR! write after end
npm ERR! write after end
```

File: src/gm3/components/catalog/legend.js (hunk 1 above)
Comment (theduckylittle) [reply src/gm3/components/catalog/legend.js]: "%s/"/'/" has caught the license text.

File: src/gm3/actions/mapSource.js (hunk 2 above)
Comment (theduckylittle) [reply src/gm3/actions/mapSource.js]: I'm not sure it did! Or it worked due to a scope issue with `var`

**Comment by theduckylittle** [comment]
What version of node / npm? I'm not having any issues.

**Comment by klassenjs** [comment]
v8.9.4 and v5.7.1

**Comment by klassenjs** [comment]
grunt build fails

```
Running "eslint:target" (eslint) task
Warning: Cannot find module 'eslint-config-react-app'
Referenced from: /home/jimk/Public/GeoMoose/gm3/eslint.config.js Use --force to continue.

Aborted due to warnings.
```

**Comment by theduckylittle** [comment]
Some stuff is missing from the package.json... I'll need to play catch up with it

**Comment by klassenjs** [comment]
Also noticed `npm audit` has some suggestions.

**Comment by klassenjs** [comment]
Did you know canvas made it into package.json?

**Comment by klassenjs** [comment]
I am getting eslint errors:

```
$ grunt build
Running "lintless:all" (lintless) task
Linting...OK

Running "eslint:target" (eslint) task

/home/jimk/Public/GeoMoose/gm3/src/index.js
  60:1  warning  Unexpected var, use let or const instead  no-var
  75:1  warning  Unexpected var, use let or const instead  no-var

/home/jimk/Public/GeoMoose/gm3/src/services/findme.js
  60:9   warning  Unexpected var, use let or const instead    no-var
  61:9   warning  Unexpected var, use let or const instead    no-var
  62:9   warning  Unexpected var, use let or const instead    no-var
  62:13  warning  'coord' is assigned a value but never used  no-unused-vars
  65:9   warning  Unexpected var, use let or const instead    no-var
  77:9   warning  Unexpected var, use let or const instead    no-var
  77:27  error    'gm3' is not defined                        no-undef
  84:9   warning  Unexpected var, use let or const instead    no-var
  85:9   warning  Unexpected var, use let or const instead    no-var

/home/jimk/Public/GeoMoose/gm3/src/services/geocode-bing.js
  56:9   warning  Unexpected var, use let or const instead  no-var
  68:9   warning  Unexpected var, use let or const instead  no-var
  70:9   warning  Unexpected var, use let or const instead  no-var
  71:9   error    'gm3' is not defined                      no-undef
  96:17  warning  Unexpected var, use let or const instead  no-var
  97:17  warning  Unexpected var, use let or const instead  no-var
  98:21  warning  Unexpected var, use let or const instead  no-var
  99:21  warning  Unexpected var, use let or const instead  no-var

/home/jimk/Public/GeoMoose/gm3/src/services/geocode-osm.js
  59:9   warning  Unexpected var, use let or const instead  no-var
  71:9   warning  Unexpected var, use let or const instead  no-var
  73:9   warning  Unexpected var, use let or const instead  no-var
  74:9   warning  Unexpected var, use let or const instead  no-var
  75:9   error    'gm3' is not defined                      no-undef
  84:17  warning  Unexpected var, use let or const instead  no-var
  85:21  warning  Unexpected var, use let or const instead  no-var
  86:21  warning  Unexpected var, use let or const instead  no-var
  98:28  error    'gm3' is not defined                      no-undef

/home/jimk/Public/GeoMoose/gm3/src/services/identify.js
  71:9   warning  Unexpected var, use let or const instead  no-var
  88:9   warning  Unexpected var, use let or const instead  no-var
  90:13  warning  Unexpected var, use let or const instead  no-var
  92:13  warning  Unexpected var, use let or const instead  no-var

/home/jimk/Public/GeoMoose/gm3/src/services/search.js
   68:9   warning  Unexpected var, use let or const instead  no-var
   69:13  warning  Unexpected var, use let or const instead  no-var
  113:9   warning  Unexpected var, use let or const instead  no-var
  115:13  warning  Unexpected var, use let or const instead  no-var
  117:13  warning  Unexpected var, use let or const instead  no-var
  158:9   warning  Unexpected var, use let or const instead  no-var
  159:13  warning  Unexpected var, use let or const instead  no-var
  160:13  warning  Unexpected var, use let or const instead  no-var

/home/jimk/Public/GeoMoose/gm3/src/services/select.js
  105:13  warning  Unexpected var, use let or const instead  no-var
  106:13  warning  Unexpected var, use let or const instead  no-var
  107:13  warning  Unexpected var, use let or const instead  no-var
  114:13  warning  Unexpected var, use let or const instead  no-var
  116:13  warning  Unexpected var, use let or const instead  no-var
  133:9   warning  Unexpected var, use let or const instead  no-var
  137:13  warning  Unexpected var, use let or const instead  no-var
  139:13  warning  Unexpected var, use let or const instead  no-var

/home/jimk/Public/GeoMoose/gm3/src/tasks/grunt-less-linter.js
  33:7   warning  'fs' is assigned a value but never used         no-unused-vars
  45:17  warning  'chr' is never reassigned. Use 'const' instead  prefer-const
  77:9   warning  Unexpected var, use let or const instead        no-var

✖ 51 problems (4 errors, 47 warnings)
  0 errors, 45 warnings potentially fixable with the `--fix` option.

Warning: Task "eslint:target" failed. Use --force to continue.

Aborted due to warnings.
```

**Comment by theduckylittle** [comment]
Try `npm run build`.

**Comment by klassenjs** [comment]
```
$ npm run build
npm ERR! missing script: build

npm ERR! A complete log of this run can be found in:
npm ERR!     .../.cache/npm/_logs/2018-09-20T16_32_00_109Z-debug.log
```

**Comment by theduckylittle** [comment]
Serves me right... I'm working on fixing the linting issue right now. I see that I have not moved the build script into package.json yet.

**Comment by theduckylittle** [comment]
I realized I forgot to lint the tests as well.

**Comment by klassenjs** [comment]
It passes the tests and builds now.  It seems to work, but haven't tested extensively.

**Comment by klassenjs** [comment]
One test failed:

```
 FAIL  tests/gm3/components/toolbar.test.js
  ● Test suite failed to run

    Cannot find module './button' from 'index.js'

    
      
      at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:169:17)
      at Object.<anonymous> (src/gm3/components/toolbar/index.js:223:41)

```

**Comment by klassenjs** [comment]
Looks like the build fails on missing button also.

**Comment by theduckylittle** [comment]
Rookie mistake! Forgot to add in two new files.

**Comment by klassenjs** [comment]
The "circle with check mark" in the catalog/visible tabs is confusing to me.  Looks like a radio-button, acts like a check-box.

The Measure tool doesn't work.  Super Tab is blank.

**Comment by klassenjs** [comment]
Otherwise looks good.

**Comment by klassenjs** [comment]
Looks good to me.

**Pull request merged**